app/src/main/java/com/github/sdp/tarjetakuna/ui/scanner/ScannerFragment.kt

```diff
@@ -20,41 +36,276 @@ class ScannerFragment : Fragment() {
-            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
-                scannerViewModel.setTextInformation(it.resultCode)
-                binding.scannerImage.setImageBitmap(
-                    Compatibility.getDataActivityResult(
-                        it,
-                        "data",
-                        Bitmap::class.java
-                    )
+        // Code adapted from https://developer.android.com/training/camerax
+        // Request camera permissions
+        if (allPermissionsGranted()) {
+            startCamera()
+        } else {
+            this.activity?.let {
+                ActivityCompat.requestPermissions(
+                    it,
+                    REQUIRED_PERMISSIONS,
+                    REQUEST_CODE_PERMISSIONS
                 )
             }
+        }
+
+        // bind the view model to the layout when text or object is detected in image
+        scannerViewModel.textDetected.observe(viewLifecycleOwner) {
+            binding.scannerTextInImageText.text = it.text
+        }
+
+        scannerViewModel.objectDetected.observe(viewLifecycleOwner) {
+            binding.scannerObjectInImageText.text = it.text
+        }
 
         binding.scannerScanButton.setOnClickListener {
-            registerImage.launch(Intent(MediaStore.ACTION_IMAGE_CAPTURE))
+            takePhoto()
         }
 
+        cameraExecutor = Executors.newSingleThreadExecutor()
+
         return root
     }
 
+    // Code adapted from https://developer.android.com/training/camerax
+    /**
+     * Check if all permissions are granted or not
+     */
+    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all {
+        val baseContext = this.context ?: return false
+        ContextCompat.checkSelfPermission(baseContext, it) == PackageManager.PERMISSION_GRANTED
+    }
+
+    /**
+     * Check whether permissions have been granted after the request,
+     * replaces the onRequestPermissionsResult method from the activity
+     */
+    fun requestPermissionsResult(requestCode: Int, perm: Array<String>, results: IntArray) {
+        // check that permissions results is for scanner fragment
+        if (requestCode == REQUEST_CODE_PERMISSIONS) {
+            // check if all permissions are granted
+            if (allPermissionsGranted()) {
+                startCamera()
+            } else {
+                // if not, show a toast and navigate back out of this fragment
+                Toast.makeText(
+                    this.context,
+                    "Permissions not granted by the user.",
+                    Toast.LENGTH_SHORT
+                ).show()
+                // TODO : should be independent of MainActivity
+                (this.activity as MainActivity).navigateUp()
+            }
+        }
+    }
+
+    /**
+     * Take a photo and save it to the device
+     */
+    private fun takePhoto() {
+        // allow only one photo to be taken at a time
```

Comment:
if you double click on the scan button fast, it will trigger 2 pictures at the same time, before the first one is saved.
It was actually not a problem, but i was trying to find a way to know when a picture was being in process of being taken and saved to wait until it's done. But not needed anymore, i could remove it if needed